File: 프로그래머스/lv3/김유현/부대복귀.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

vector<int> adj[100004];

int visited[100004];

vector<int> solution(int n, vector<vector<int>> roads, vector<int> sources, int destination) {
    vector<int> answer;
    
    for(vector<int> road : roads) {
        int a = road[0], b = road[1];
        
        adj[a].push_back(b);
        adj[b].push_back(a);
    }
        
    // destination에서 갈 수 있는 모든 곳 bfs로 구하기
    queue<int> q;
    q.push(destination);
    visited[destination] = 1;

    while(!q.empty()) {
        int curr = q.front();
        q.pop();
        
        for(int next : adj[curr]) {
            if (visited[next])
                continue; 
            
            visited[next] = visited[curr] + 1;
            q.push(next); 
        }
        
    }
    
    for(int source : sources) {
        answer.push_back(visited[source]-1);
    }
    
    return answer;
}

File: 프로그래머스/lv3/이선영/부대복귀.java
import java.util.*;

class Solution {
    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        int[] answer = new int[sources.length];
        int[] min = new int[n+1];
        List<Integer>[] list = new ArrayList[n+1];
        
        for(int i=0; i<list.length; i++){
            list[i] = new ArrayList<>();
        }
        
        for(int i=0; i<roads.length; i++){
            int[] road = roads[i];
            list[road[0]].add(road[1]);
            list[road[1]].add(road[0]);
        }
        
        boolean[] visited = new boolean[n+1];
        
        Arrays.fill(min, -1);

        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{destination, 0});
        
        
        while(!q.isEmpty()){
            int[] now = q.poll();
            if(visited[now[0]]) continue;
            
            visited[now[0]] = true;
            min[now[0]] = now[1];
            
            for(int j=0; j<list[now[0]].size(); j++){
                int next = list[now[0]].get(j);
                q.offer(new int[]{next, now[1]+1});
            }   
        }
        
        
        int tmp = 0;
        for(int i: sources){
            answer[tmp++] = min[i];
        }
        
        
        return answer;
    }
    
   
}
